### PR TITLE
Fog and mist weather

### PR DESCRIPTION
**Pull request by GuardianDll**
#### Summary
Content "add new weather: mist and fog. maybe."
#### Purpose of change
I want to make a fog
#### Describe the solution
Currently fog and mist require extremely high humidity, more than 95%, small wind, and some level of temperature drop
Also @andrei8l fixed autotests, that require clean vision to pass, 
#### Testing
~~And here is the problem - i can't properly test this solution, because such high humidity happen only when it rains. Also i am not 100% sure i get how it should be done - from what i know there should be also a dew point calculation, but i still don't understand how it should be utilized, as when relative humidity is 100%, the dew point is equal the temperature itself~~ yep, now everything works as expected
image

### CODE REVIEW

**Comment by NetSysFire** [comment]
irl fog can also happen when the air temperature is lower than the water in a nearby body of water e.g river, which happens frequently where I live. I saw some discussion on discord that specific overmap tiles can cause specific weather, but that is unfortunately out of scope right now.

**Comment by GuardianDll** [comment]
Found some info, that to generate the mist, the temperature do not need to be same as dew point (every source i was able to find says it should be lower, but it's like impossible physically), but close to it at least for 2.5 centigrade.
With new `math`, i plan to use a [Tetens formula](https://en.wikipedia.org/wiki/Tetens_equation), which calculate a dew point, and another math will calculate a dew point factor - if the factor is bigger than 0, it can generate a mist, and if it's bigger than 1, it can create a fog

```json

{ "math": [ "dew_point", "=", "(237.7 * log(u_val('weather: humidity')/100) + 17.27 * u_val('weather: temperature')) / (17.27 - 237.7 * log(u_val('weather: humidity')/100))" ] },
{ "math": [ "dew_point_factor", "=", "u_val('weather: temperature') - (dew_point + 2.5)" ] }

```

Thoughts?

**Comment by andrei8l** [comment]
Using [this](https://journals.ametsoc.org/view/journals/bams/86/2/bams-86-2-225.xml?tab_body=pdf) as a reference, the formula for dew point temperature should be
```JSON
{
  "math": [
    "dprhx",
    "=",
    "log( u_val('weather:humidity') / 100 ) + 17.625 * u_val('weather:temperature') / ( 243.04 + u_val('weather:temperature') )"
  ]
},
{ "math": [ "dew_point", "=", "243.04 * dprhx / ( 17.625 - dprhx )" ] }
```
But the `"weather": "temperature"` function returns temperature in Fahrenheit (undocumented) and the formula needs it in Celsius so you also need to convert it. I'll port the weather functions after [65058](https://github.com/CleverRaven/Cataclysm-DDA/pull/65058/files) and add conversion functions too.

In the linked paper there are also several simpler formulas that might be just as suitable here.

**Comment by GuardianDll** [comment]
Ok, i can't actually test it occur naturally, but Ramza says weather code is pretty obscure and may not work properly, so i just make how it should be with hope someone refactor the weather and fix this in some future

File: data/json/weather_type.json [hunk 1]
```diff
@@ -269,6 +269,76 @@
...
+    "sound_attn": 0,
+    "dangerous": false,
+    "precip": "light",
+    "rains": false,
+    "sound_category": "silent",
+    "priority": 40,
+    "condition": {
+      "and": [
+        { "math": [ "u_val( 'weather:windpower' )", ">", "12" ] },
+        { "math": [ "u_val( 'weather:windpower' )", "<", "0" ] },
```
Comment (andrei8l) [inline data/json/weather_type.json:292]: Isn't it supposed to be like this?
```suggestion
        { "math": [ "u_val( 'weather:windpower' )", "<", "12" ] },
        { "math": [ "u_val( 'weather:windpower' )", ">", "0" ] },
```

Comment (GuardianDll) [reply data/json/weather_type.json:292]: oh god, i really inversed the wind values and miss it

**Comment by GuardianDll** [comment]
Oh no, now the vision tests are broken

**Comment by andrei8l** [comment]
There's a genuine vision bug for weathers with `sight_penalty` not `1.0`.

[output.webm](https://user-images.githubusercontent.com/68240139/232167659-e5657ecc-3582-4f35-85b0-d6aee212141a.webm)

but that needs to be addressed elsewhere.


<details>
<summary>This should fix the vision test for now</summary>

```diff
diff --git a/tests/vision_test.cpp b/tests/vision_test.cpp
index 51c1ecd30c..e5239c6f64 100644
--- a/tests/vision_test.cpp
+++ b/tests/vision_test.cpp
@@ -149,7 +149,6 @@ struct vision_test_case {
                       const time_point &time ) : setup( setup ), expected_results( expectedResults ), time( time ) {}
 
     void test_all() const {
-        scoped_weather_override weather_clear( WEATHER_CLEAR );
         Character &player_character = get_player_character();
         g->place_player( tripoint( 60, 60, 0 ) );
         player_character.worn.clear(); // Remove any light-emitting clothing
@@ -159,6 +158,7 @@ struct vision_test_case {
         player_character.clear_moncams();
         clear_map( -2, OVERMAP_HEIGHT );
         g->reset_light_level();
+        scoped_weather_override weather_clear( WEATHER_CLEAR );
 
         REQUIRE( !player_character.is_blind() );
         REQUIRE( !player_character.in_sleep_state() );
@@ -541,11 +541,11 @@ TEST_CASE( "vision_junction_reciprocity", "[vision][reciprocity]" )
         player_in_junction ?
         std::vector<std::string>{
             "u#  ",
-            "-- Z",
+            "---Z",
 }:
         std::vector<std::string>{
             "z#  ",
-            "-- u",
+            "---u",
         },
         player_in_junction ?
         std::vector<std::string>{
```
</details>
I'll take another look tomorrow if it still fails.

**Comment by andrei8l** [comment]
I haven't been able to figure out the shadowcasting bug. We can work around it for the test unit by overspecifying the setup though
<details>
<summary>This will fix `vision_junction_reciprocity` or your money back</summary>

```diff
diff --git a/tests/vision_test.cpp b/tests/vision_test.cpp
index e5239c6f64..f7caaceec1 100644
--- a/tests/vision_test.cpp
+++ b/tests/vision_test.cpp
@@ -536,25 +536,34 @@ TEST_CASE( "vision_single_tile_skylight", "[shadowcasting][vision]" )
 TEST_CASE( "vision_junction_reciprocity", "[vision][reciprocity]" )
 {
     bool player_in_junction = GENERATE( true, false );
+    CAPTURE( player_in_junction );
 
     vision_test_case t {
         player_in_junction ?
         std::vector<std::string>{
-            "u#  ",
-            "---Z",
+            "###   ",
+            "#u####",
+            "#---z#",
+            "######",
 }:
         std::vector<std::string>{
-            "z#  ",
-            "---u",
+            "###   ",
+            "#z####",
+            "#---u#",
+            "######",
         },
         player_in_junction ?
         std::vector<std::string>{
-            "4466",
-            "4446",
+            "444666",
+            "444666",
+            "444466",
+            "444466",
 }:
         std::vector<std::string>{
-            "4444",
-            "4444",
+            "666666",
+            "444444",
+            "444444",
+            "444444",
         },
         day_time
     };
@@ -562,14 +571,14 @@ TEST_CASE( "vision_junction_reciprocity", "[vision][reciprocity]" )
     monster *zombie = nullptr;
     tile_predicate spawn_zombie = [&]( map_test_case::tile tile ) {
         zombie = g->place_critter_at( mon_zombie, tile.p );
+        get_map().ter_set( tile.p + tripoint_above, ter_t_flat_roof );
         return true;
     };
 
     t.set_up_tiles =
-        ifchar( 'C', spawn_moncam ) ||
-        ifchar( 'Z', spawn_zombie ) ||
         ifchar( 'z', spawn_zombie ) ||
         t.set_up_tiles;
+    t.flags.headlamp = true;
     t.test_all();
 
     if( player_in_junction ) {
```

</details>

There's also a failure in the `weather realism` test. Wikipedia says fog and mist aren't precipitation
<details>
<summary>Potential realism fix; haven't tested in game whether fog/mist still occur</summary>

```diff
diff --git a/data/json/weather_type.json b/data/json/weather_type.json
index a948fd8ca4..df4aea49bd 100644
--- a/data/json/weather_type.json
+++ b/data/json/weather_type.json
@@ -282,10 +282,11 @@
     "sun_multiplier": 0.3,
     "sound_attn": 0,
     "dangerous": false,
-    "precip": "light",
+    "precip": "none",
     "rains": false,
     "sound_category": "silent",
     "priority": 20,
+    "required_weathers": [ "clear", "sunny" ],
     "condition": {
       "and": [
         { "math": [ "u_val( 'weather:windpower' )", "<", "12" ] },
@@ -307,7 +308,7 @@
     "sun_multiplier": 0.3,
     "sound_attn": 0,
     "dangerous": false,
-    "precip": "heavy",
+    "precip": "none",
     "rains": false,
     "sound_category": "silent",
     "priority": 20,
```
</details>

Sorry it took me so long to get back to this.

**Comment by GuardianDll** [comment]
\> Sorry it took me so long to get back to this

Nothing to worry about!

\> Wikipedia says fog and mist aren't precipitation

i was not sure for what `precip` is used for (`Amount of associated precipitation` from doc doesn't help much), so i just cargo culted it

**Comment by GuardianDll** [comment]
ok i somewhat feel that sunny weather won't work, because it require humidity less than 70% - totally opposite of ~100% for fog forming. Maybe replace it with "cloudy"?

**Comment by bombasticSlacks** [comment]
I'm not exactly sure where people sit on this, is this good to merge or is there still things to be resolved?

**Comment by GuardianDll** [comment]
Tests looks good, in game it seems okay, so i think it good to go

**Pull request merged**